src/title.js

```diff
@@ -0,0 +1,441 @@
+	 *
+	 * @private
+	 * @returns {module:engine/model/element~Element|undefined}
+	 */
+	_getTitleElement() {
+		const root = this.editor.model.document.getRoot();
+
+		for ( const child of root.getChildren() ) {
+			if ( child.is( 'title' ) ) {
+				return child;
+			}
+		}
+	}
+
+	/**
+	 * Model post-fixer callback that ensures `title` has only one `title-content` child.
+	 * All additional children should be moved after `title` element and renamed to a paragraph.
+	 *
+	 * @private
+	 * @param {module:engine/model/writer~Writer} writer
+	 * @returns {Boolean}
+	 */
+	_fixTitleContent( writer ) {
+		const title = this._getTitleElement();
+
+		if ( !title ) {
+			return false;
+		}
+
+		const titleChildren = Array.from( title.getChildren() );
+		let hasChanged = false;
+
+		// Skip first child because it is an allowed element.
+		titleChildren.shift();
+
+		for ( const titleChild of titleChildren ) {
+			writer.move( writer.createRangeOn( titleChild ), title, 'after' );
+			writer.rename( titleChild, 'paragraph' );
+			hasChanged = true;
+		}
+
+		return hasChanged;
+	}
+
+	/**
+	 * Model post-fixer callback that takes care about creating and keeping `title` element as a first child in a root.
+	 *
+	 * @private
+	 * @param {module:engine/model/writer~Writer} writer
+	 * @returns {Boolean}
+	 */
+	_fixTitleElement( writer ) {
+		const model = this.editor.model;
+		const modelRoot = model.document.getRoot();
+		let hasChanged = false;
+		let numberOfMovedElements = 0;
+
+		// Loop through root children and take care about a proper position of a title element.
+		// Title always has to be the first element in the root.
+		for ( const rootChild of Array.from( modelRoot.getChildren() ) ) {
+			// If the first element is not a title we need to fix it.
+			if ( rootChild.index === 0 && !rootChild.is( 'title' ) ) {
+				const title = this._getTitleElement();
+
+				// Change first element to the title if it can be a title.
+				if ( Title.titleLikeElements.has( rootChild.name ) ) {
+					const position = model.createPositionBefore( rootChild );
+					const title = writer.createElement( 'title' );
+
+					writer.insert( title, position );
+					writer.append( rootChild, title );
+					writer.rename( rootChild, 'title-content' );
+
+					// After changing element to the title it has to be filtered out from disallowed attributes.
+					model.schema.removeDisallowedAttributes( [ rootChild ], writer );
+
+				// If the first element cannot be a title but title is already in the root
+				// than move the first element after a title.
+				} else if ( title ) {
+					const positionAfterTitle = writer.createPositionAt( title, 'after' );
+
+					// To preserve correct order when more than one element is moved in the one post-fixer call.
+					const targetPosition = positionAfterTitle.getShiftedBy( numberOfMovedElements );
+
+					writer.move( writer.createRangeOn( rootChild ), targetPosition );
+					numberOfMovedElements++;
```

Comment:
I find this not appealing :) Maybe this whole logic could be re-structured to a check title condition and fix then rename remaining titles? We can get the first child and an existing title fast. We can then either rename the first element to a title or move the existing title at the begging of the root.

We do not have to do everything in one loop IMHO.

Reply:
ps.: this might invalidate my comment about the extracting this to separate methods.